### PR TITLE
Zr 354 classname updates

### PR DESCRIPTION
**Pull request by lancetipton**
**Ticket [ZEN-354](https://jira.simpleviewtools.com/browse/ZEN-354)** 

**Notes**
* Originally this ticket was to add a `dataSet.class` attribute to the components
  * This has been changed to instead use the `className` prop
  * This in more consistent with other web platforms
  * On Native, it is ignored

## Context
* Events-Force is defining their styles in CSS
* Their clients can define custom CSS as well
* Currently, the tap, using re-theme, is only compatible with JSON style objects
* We need to be able to pull styles from stylesheets already applied to the DOM and apply them to keg-components

## Goal
* Update all keg-components to accept a className prop
  * Using the `useClassname` and `useClasslist` hooks, the `className` can be used instead of the `dataSet`
* Add a default `className` prop to all primitive components
* Merge the passed in `className` prop with the default `className` prop

## Updates
**Global**
  * Removed `@expo/vector-icons` && `react-native-vector-icons`
  * Added `react-native-svg` && `react-native-svg-web`
  * Added `svgr cli` to convert SVG's to React Native Components
  * Added scripts helpers to do SVG conversion
  * Removed all dataSet references
  * Updated alias / extentions to work like Re-Theme does
* `repos/keg-components/README.md`
  * Updated with more information
  * Added notes about updated / added code
* `repos/keg-components/configs/aliases.json`
  * Cleaned up to match code updates
* `repos/keg-components/configs/jest.config.js`
  * Removed `@expo/vector-icons`
* `repos/keg-components/setupTests.js`
  * Added mocks for `react-native-svg` to `react-native-svg-web`
  * Removed a number of the `<component>.wrapper.js ` as they are no longer needed

**Assets/Icons**
* `repos/keg-components/src/assets/icons/**`
  * Added a number of svg icons as react components
    * Mostly used in stories, but also used in some places of the code
* `repos/keg-components/src/assets/icons/svgIcon/svgIcon.js`
  * Added SVG component for consistent SVG icon rendering
* `repos/keg-components/src/assets/spinners/**`
  * Removed spinners as they are no longer being used

**Components**
* `repos/keg-components/src/components/**`
  * Updated to use `useClassList` and `useClassName` hooks as needed
  * Updated to use `SVG` component as needed
  * Added a **Touchable** component
    * Replaced components that were using `TouchableOpacity` and `TouchableNativeFeedback`
  * Added **accessibilityRole** to components where needed
    * This helps with symantec HTML, as well Dom element type to be used 
  * Removed Component wrapper in cases where no longer needed
    * `Button` / `Form` components
  * Added `createElement.js` component
    * Allows creating custom Components using the `react-native-web` framework
  * Fixed `Checkbox` component styles
    * Added styles are defined here so that they are enforced
    * Due to how the checkbox is designed, these styles should not be changed 
    * Which is why they are not defined within the theme
  * Added helpers to `noop`
    * `noPropObj` && `noPropArr`
    * Replace locations in components where they were just being set to an empty object or array
  * Fixed bug with TouchableIcon causing the wrapper styles to not be used
  * Updated `style` prop in a number of locations to use the `array` syntax
  * `repos/keg-components/src/components/box/textBox.js`
    * Fixed issue with icon styles to being applied properly
* `repos/keg-components/src/components/card/card.js`
  * Lots of updates to clean this up
  * Added more stories
* `repos/keg-components/src/components/drawer/drawer.js`
  * Added this but have not full set it up yet
  * I'm using it in one of my own taps, so figured I copy it over to here
  * I'll updating it more in the next pull request
  * It's not being exported, so does not effect anything
* `repos/keg-components/src/components/form/select/picker.js`
  * Same as the `drawer` component
  * Needs a lot more work, but this is a start
  * It's not being exported, so does not effect anything
* `repos/keg-components/src/components/image/image.js`
  * Lots of clean up to fix how Images and loading works
    * Now uses the Loading component properly
    * Added some custom code here to ensure the classes get applied to the image elements
* `repos/keg-components/src/components/loading/loading.js`
  * Lots of clean up to fix how loading and spinners works
    * Spinners are now centered, and look much better

**Hooks** - Added a number of hooks for the className updates
* `repos/keg-components/src/hooks/useClassList.js`
  * Joins an array of class names together and memoizes them
* `repos/keg-components/src/hooks/useClassName.js`
  * Returns a function that should be set as the element ref
  * When the ref gets called, it gets the element and applies the `classNames` to the dom element
* `repos/keg-components/src/hooks/useThemePath.js`
  * Cleaned this up a lot
  * Now easier to understand, and seems to be faster
  * It's also doing a lot less
* `repos/keg-components/src/hooks/useThemeTypeAsClass.js`
  * Helper hook to use the `themePath`, and `type` props of a component to build the className
  * Calls `useClassList` internally

**Theme**
  * `repos/keg-components/src/theme/themeDefaults.js`
    * Updated theme defaults to allow being overwritten
    * Updated all references to `defaults.json` to use `getThemeDefaults` instead 
    * Added a `setThemeDefaults` that accepts a `defaults` object
    * **IMPORTANT** - This should be called prior to theme initialization
      * This way the defaults will be updated before the theme is built
  * `repos/keg-components/src/utils/styles/buildColorStyles.js`
    * Added `buildSurfaceStyles` helper to build styles based on defined `surfaces`
  * `repos/keg-components/src/theme/components/loading.js`
    * Updated how the styles are created
    * Uses the `buildSurfaceStyles`
  * 
**Keg-Core**
* `app.json`
  * Added mapping of `react-native-svg` to `react-native-svg-web` on web platform

## Testing
* Run `keg dpg run package=docker.pkg.github.com/simpleviewinc/keg-packages/keg-components:zr-354-classname-updates`
  * All components should have a `keg-*` class name applied to them
  * The `*` is based on the component name and location
  * So `keg-card-container` would be the container component with-in the card

**Steps to Test**
  * Open up the inspector in your browser
  * Then, starting with the first component; right click it an select `inspect`
  * Find the component within the `inspectors` dom tree view
  * Ensure that it has a class name applied to it
    * It should look like  [this](http://snpy.in/Y1WwtP)
  * Do this for all components and ensure they have a `keg-*` class applied to them
  * **IMPORTANT** - There are some cases `react-native-web` will generate child dom elements
    * These will not have the class applied to them
    * Technically I could have added them, doing something like what was done on the Image component
    * But it's possible for `react-native-web` to change what elements are created in future versions, which would break how this works.
    * I did do it for the image, because I felt it was important to have a class directly applied to images
      * Images tend to have more styles applied to them, which means they need a way to be accessed
      * You can see an example of the `react-native-web` generated elements [here](http://snpy.in/1IGHBI)
        * In that image, the element with the `keg-image-sibling` class is created by `react-native-web`; not `keg-components`

### CODE REVIEW

**Review by mpcarolin — CHANGES_REQUESTED** [review]
Overall looks good, but I have a few thoughts on the code, and I encountered a couple of issues. The issues may or may not be related to these changes. If not, let me know, and I'll approve this. `useClassName` is great though and I'm excited to try all of this out in the tap.

Issues

1. The performance of clicking the `button` and `touchable` components, or anything else that is touchable, seems to be pretty bad. If I rapidly click it 4 or 5 times, for example, it lags significantly. 

2.  The first two modal stories (One at a time, multiple modals), do not seem to be doing anything for me (no modals appear). The second two work as expected.

3. The `disabled` behavior in `select` no longer works, at least in the story.

File: repos/keg-components/src/hooks/useClassList.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * Builds an array of classNames, memoizes then returns them
+ * <br/>Returned function updates the element to include a className
+ * <br/>Class Name can from from the className prop, and the defClass value
+ * @param {Array|string} className - Array of classes to add
+ * @param {Array} classList - Array of default classes
+ *
+ * @returns {function} - Ref function to be added to the component
+ */
+export const useClassList = (className, classList = noPropArr) => {
+  const classListArr = isArr(classList) ? classList : [classList]
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/hooks/useClassList.js:15]: Nice, totally forgot about that one; updated

File: repos/keg-components/src/hooks/useClassName.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * Returns a function that should be set as the element ref
+ * <br/>Returned function updates the element to include a className
+ * <br/>Class Name can from from the className prop, and the defClass value
+ * @param {Array[string]} defClass - Default class to add to the component
+ * @param {string|Array[string]} className - Class or an array of classes to add to the element
+ * @param {Object|function} ref - Ref object passed to the consuming component
+ *
+ * @returns {function} - Ref function to be added to the component
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/hooks/useClassName.js:14]: No prob, updated

File: repos/keg-components/src/components/touchable/touchable.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * @property {Function} props.onPress - function to do when button is pressed
+ * @property {Object} props.children - Child components of the component
+ * @property {Object} props.ref - reference to native element
+ *
+ */
+export const Touchable = React.forwardRef((props, ref) => {
+  const { className, showFeedback = true, ...attrs } = props
+  const Component = showFeedback ? TouchableComp : TouchableWithoutFeedback
+
+  const touchRef = useClassName('keg-touchable', className, ref)
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/components/touchable/touchable.js:30]: sure, updated

File: repos/keg-components/src/hooks/useThemeTypeAsClass.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import { colors } from '../theme/colors'
+import { useClassList } from 'KegClassList'
+import { noOpObj } from '../utils/helpers/noop'
+
+/**
+ * Uses the surfaces defined in the colors to build surface specific classes for
+ * @param {string} themeRef - Either the type or themePath passed to the component
+ * @param {string} defRef - The default reference to use for the class
+ *
+ * @returns {Array} - Built classList based on surfaces and themeRef
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/hooks/useThemeTypeAsClass.js:12]: Updated to be `Array<string>` and fixed the where it's returning a string. It should never return a string. It should always return an Array

File: repos/keg-components/src/hooks/useThemeTypeAsClass.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import { useMemo } from 'react'
+import { get, isArr } from '@keg-hub/jsutils'
+import { colors } from '../theme/colors'
+import { useClassList } from 'KegClassList'
+import { noOpObj } from '../utils/helpers/noop'
+
+/**
+ * Uses the surfaces defined in the colors to build surface specific classes for
+ * @param {string} themeRef - Either the type or themePath passed to the component
+ * @param {string} defRef - The default reference to use for the class
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/hooks/useThemeTypeAsClass.js:10]: Updated

File: repos/keg-components/src/components/form/select/picker.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  Modal,
+  Text,
+  TouchableHighlight,
+  View,
+  StyleSheet,
+  FlatList,
+} from 'react-native'
+import PropTypes from 'prop-types'
+
+export default class ModalPicker extends Component {
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/components/form/select/picker.js:12]: Good Idea, updated

**Review by daniel112 — CHANGES_REQUESTED** [review]
looks great, the classnames appear where they should be. 
there's just some minor things that I encountered

File: repos/keg-components/src/assets/icons/svgIcon/svgIcon.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    size,
+    stroke,
+    style = noPropObj,
+    viewBox,
+    ...attrs
+  } = props
+
+  const theme = useTheme()
+  const strokeColor = stroke || border || style.border || color || style.color
+  const fillColor = fill || color || style.color || stroke
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/assets/icons/svgIcon/svgIcon.js:44]: Updated

File: repos/keg-components/README.md [hunk 8]
```diff
@@ -113,4 +121,43 @@ export const button = {
...
 }
 ```
-* in the first case, the theme object would not have the `content` styles on the `$web` platform, making it inconsistent with `$native`
+* In the first case, the theme object would not have the `content` styles on the `$web` platform, making it inconsistent with `$native`
+
+
+## Icons
+ * **TODO - add info on adding SVG's**
+
+## Accessability
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/README.md:130]: Updated

File: repos/keg-components/.storybook/preview.js [hunk 9]
```diff
@@ -1,20 +1,47 @@
...
+          const isCompName = component.__docgenInfo.displayName.toLowerCase().trim() === part.toLowerCase().trim()
+          return isCompName || !part.trim() ? full : parsePart(full, part)
+        }, '')
+        return `${parsed}<br/>`
+      }
+      return notes
+        ? typeof notes === 'string' ? notes : notes.markdown || notes.text
+        : null
+    },
+  },
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/.storybook/preview.js:34]: Updated

File: repos/keg-components/src/components/box/textBox.stories.js [hunk 10]
```diff
@@ -26,7 +26,7 @@ const TextBoxStory = ({ type, useClipboard }) => {
   )
 }
 
-storiesOf('Box | TextBox', module)
+storiesOf('Box/TextBox', module)
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/components/box/textBox.stories.js:29]: This is fixed

File: repos/keg-components/src/components/createElement/createElement.js [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import { unstable_createElement as RNCreateElement } from 'react-native'
+
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/components/createElement/createElement.js:2]: This is a work in progress. I'll update it soon. For now I added some notes about it being WIP. 
Feel free to use it, but I'm not really considering it supported at this point. It needs a few things first

File: repos/keg-components/src/theme/themeDefaults.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import { deepMerge } from '@keg-hub/jsutils'
+import defaults from './defaults.json'
+
+let __themeDefaults = defaults
+
+export const setThemeDefaults = overrides => {
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/theme/themeDefaults.js:6]: This also is a work in progress. I haven't tested any of this, which is why it's not exported. I'll add a not at the top

**Review by mpcarolin — APPROVED** [review]
Looks good to me 👍. Passed all testing, and the buttons no longer lag.

**Review by daniel112 — APPROVED** [review]
looks gucci

**Pull request merged**